common/model/transfer.c

```diff
@@ -1,676 +1,474 @@
+      current_index++;
+    }
+  }
+  // Squeeze kerl to get the bundle hash
+  kerl_squeeze(kerl, bundle_trit, HASH_LENGTH_TRIT);
+  flex_trits_from_trits(transfer_ctx->bundle, HASH_LENGTH_TRIT, bundle_trit,
+                        HASH_LENGTH_TRIT, HASH_LENGTH_TRIT);
+}
+
+transfer_iterator_t* transfer_iterator_new(transfer_t* transfers[], size_t len,
+                                           Kerl* kerl) {
+  transfer_ctx_t* transfer_ctx = NULL;
+  transfer_iterator_t* transfer_iterator =
+      (transfer_iterator_t*)calloc(1, sizeof(transfer_iterator_t));
+  if (!transfer_iterator) {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  transfer_iterator->transfers = transfers;
+  transfer_iterator->transfers_count = len;
+  transfer_ctx = transfer_ctx_new();
+  if (!transfer_ctx) {
+    transfer_iterator_free(&transfer_iterator);
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Out of Memory.\n", __func__,
+              __LINE__);
+    return NULL;
+  }
+  if (transfer_ctx_init(transfer_ctx, transfers, len)) {
+    transfer_iterator->transactions_count = transfer_ctx->count;
+    transfer_ctx_hash(transfer_ctx, kerl, transfers, len);
+    memcpy(transfer_iterator->bundle_hash, transfer_ctx->bundle,
+           FLEX_TRIT_SIZE_243);
+  } else {
+    log_error(TRANSFER_LOGGER_ID, "[%s:%d] Invalid transfers.\n", __func__,
+              __LINE__);
+  }
+  transfer_ctx_free(transfer_ctx);
+  transfer_iterator->iota_signature_gen = iota_flex_sign_signature_gen;
+  return transfer_iterator;
+}
+
+void transfer_iterator_free(transfer_iterator_t** iter) {
+  transfer_iterator_t* iter_p = *iter;
+  if (*iter || iter) {
+    if (iter_p->dynamic_transaction && iter_p->transaction) {
+      free(iter_p->transaction);
+    }
+    if (iter_p->transaction_signature) {
+      free(iter_p->transaction_signature);
+    }
+    free(iter_p);
+    *iter = NULL;
+  }
+}
+
+iota_transaction_t transfer_iterator_next(
+    transfer_iterator_t* transfer_iterator) {
+  iota_transaction_t transaction = NULL;
+
+  if (transfer_iterator->current_transfer <
+      transfer_iterator->transfers_count) {
+    // Dynamically allocate a transaction structure if none was
+    // previously provided with transfer_iterator_set_transaction
+    if (!transfer_iterator->transaction) {
+      transfer_iterator->transaction = transaction_new();
+      transfer_iterator->dynamic_transaction = 1;
+    }
+    transfer_t* transfer =
+        transfer_iterator->transfers[transfer_iterator->current_transfer];
+    size_t count = transfer_transactions_count(transfer);
+    if (transfer_iterator->current_transfer_transaction_index >= count) {
+      transfer_iterator->current_transfer++;
+      transfer_iterator->current_transfer_transaction_index = 0;
+      return transfer_iterator_next(transfer_iterator);
+    }
+    transaction = transfer_iterator->transaction;
```

Comment:
could you export that logic to a separate function?

Reply:
I think this logic is in here only.

Reply:
yeah, it is not crucial, but it makes the code clearer

Reply:
Why do you think it's not clear? I mean no complex logic and it has some comments here.